[BP-727][TEC-759] AWS Terraform Example (#20)
Bodo-inc/Bodo-examples: terraform/aws/README.md

# Spin up a cluster on AWS for Bodo using Terraform

1. [Install Terraform](https://www.terraform.io/downloads.html).

1. Create an AMI. See [this](./bodo-ami/README.md) for instructions. Skip this step if using an existing image.

1. Set up your AWS credentials. We recommend doing this through [environment variables](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html), but other methods should also work.

> 1. The provided terraform module takes the following variables:

Reviewer: We provided a terraform module example: [``bodo_cluster.sh``](./bodo_cluster.sh), which takes the following variables

Author: ```suggestion
1. The provided terraform configuration takes the following variables:
```

Author: How about this (along with the change to the last line)?

    a. ``AMI_ID`` (required): See step 2.
    
    b. ``VPC_ID`` (required): ID of the VPC you want to deploy the cluster in. You can find this value in your AWS VPC Console. Each AWS account comes with a default VPC.
    
    c. ``CLUSTER_SUBNET_ID`` (required): ID of the subnet to deploy the cluster in. You can find this value in your AWS VPC Console. Ensure that the subnet is inside the VPC defined by ``VPC_ID``.
    
    d. ``USER_IP`` (optional): Your IP address. When provided a security group rule allowing SSH access from this IP to the cluster instances is created. If using [``bodo_cluster.sh``](./bodo_cluster.sh), this is automatically set as the IP of the machine the script is being run on.

    e. ``AWS_DEFAULT_REGION`` (optional): The region to deploy the cluster in. Default: ``us-east-1``.

    f. ``CLUSTER_MEMBERS_COUNT`` (optional): Number of instances in the cluster. Default: 2.

    g. ``CLUSTER_INSTANCE_TYPE`` (optional): Instance type. See [this](https://aws.amazon.com/ec2/instance-types/) for a full list. Default: ``c5n.xlarge``.

>     We recommend using a script like [``bodo_cluster.sh``](./bodo_cluster.sh) to define these variables.

Reviewer: delete this sentence

Author: ```suggestion
    The provided [``bodo_cluster.sh``](./bodo_cluster.sh) script is a wrapper around some common terraform commands we'll use to manage our deployment. Set the above variables in this script.
```

>     We recommend using a script like [``bodo_cluster.sh``](./bodo_cluster.sh) to define these variables and run terraform.

Author: ```suggestion
    We recommend using a script like [``bodo_cluster.sh``](./bodo_cluster.sh) to define these variables.
```

1. Do a dry-run using:

        ./bodo_cluster.sh plan

    This will validate some of the arguments and output a plan of the resources that will be created. If this finishes successfully, move to the next step.

1. Create the cluster using:

        ./bodo_cluster.sh apply

    This will create the cluster and then output the public IPs of the created EC2 instances. 

>     The SSH private key for the cluster will be written to ``bodo_cluster_ssh.pem``.

Author: ```suggestion
    The SSH private key for the cluster will be written to ``bodo_cluster_ssh.pem``.
    The EC2 instances and other resources that were provisioned should be visible on your [AWS Management console](https://aws.amazon.com/console/).
```

    
1. Log into one of the EC2 instances of the cluster:

        ssh -i bodo_cluster_ssh.pem ubuntu@<PUBLIC_IP_ADDRESS>

    A machinefile for the cluster should already exist at ``/home/ubuntu/machinefile``.
    You can run code using ``mpiexec`` as usual:

        mpiexec -machinefile /home/ubuntu/machinefile -n <TOTAL_CORES_IN_CLUSTER> python ....

1. To delete the cluster, run:

>         ./bodo_cluster.sh destroy

Reviewer: Can you add a section on how to specify how many nodes of which machine you want to spin up? 
Also add a section on how to install and launch jupyter notebook/lab?

Author: > 1. how to specify how many nodes of which machine you want to spin up?

That's the `CLUSTER_MEMBERS_COUNT` variable.

> 2. how to install and launch jupyter notebook/lab?

I'll add that later. We can actually use something like https://github.com/Bodo-inc/remote_ikernel to use the cluster from a local jupyterlab server.

> 3. how to see the cluster on aws

I'll add a small line.

# Spin up a cluster on AWS for Bodo using Terraform

1. [Install Terraform](https://www.terraform.io/downloads.html).

1. Create an AMI. See [this](./bodo-ami/README.md) for instructions. Skip this step if using an existing image.

> 1. Set up your AWS credentials. We recommend doing this through [environment variables](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html), but other methods should also work.

Reviewer: This step should be done before create an AMI? (duplicate instruction in the ami readme)

Author: It's needed for both parts. If someone already has an image they might otherwise forget to set these.

> 1. We recommend using a script like [``bodo_cluster.sh``](./bodo_cluster.sh) to define variables and run terraform.

Reviewer: What's the action item here? "We recommend... " is an additional clause?

Author: I'll combine this step with the next one.

Author: Done.

> 1. The provided terraform module takes the following variables:

Reviewer: what's the command to run terraform? or is running bodo_cluster.sh runs terraform? instruction is a bit unclear here

Author: Yes, `bodo_cluster.sh` is a wrapper around the terraform executable.
